### ISSUE TITLE
RF06 Unnecessary quoted identifer ignores tsql square bracket quotes

### ISSUE THREAD

**Issue by daviewales** [sqlfluff/sqlfluff#4724]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I ran `sqlfluff lint --dialect tsql` against a tsql script with [square bracket] quoted identifiers.

e.g.

``` sql
select
    [col1],
    [col2]
from database_table
```

### Expected Behaviour

I expected this to trigger rule [RF06](https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.sphinx.Rule_RF06): Unnecessary quoted identifier.
I also hoped that `sqlfluff fix` would be able to remove the unnecessary quotes.

### Observed Behaviour

No errors or warnings were triggered.

### How to reproduce

Run `sqlfluff lint --dialect tsql` against the following:

``` sql
select
    [col1],
    [col2]
from database_table
```

### Dialect

tsql

### Version

sqlfluff, version 2.0.3

### Configuration

```
[sqlfluff]
dialect = tsql
max_line_length = 88

[sqlfluff:indentation]
allow_implicit_indents = True

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = True
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by fmms** [comment]
I have been wondering about the same issue previously as well.

Regarding the fixing, the documentation states:
> Note due to different quotes being used by different dialects supported by SQLFluff, and those quotes meaning different things in different contexts, this mode is not sqlfluff fix compatible.

However, I am not sure how big this issue is, I would expect that each dialog could define its identifier quoting style...

**Comment by daviewales** [comment]
I think the mode which is not supported is _adding_ quotes when `prefer_quoted_identifiers = True`?

Earlier in the rule I read:

> When `prefer_quoted_identifiers = False` (default behaviour), the quotes are unnecessary, except for reserved keywords and special characters in identifiers.
>
> This rule is sqlfluff fix compatible.

**Comment by alanmcruickshank** [comment]
I think it would be possible to extend RF06 to handle square quotes. To be honest, I've only just discovered, that's is valid SQL! 😄 .

Would either of you be interested in putting together a pull request? Much better that someone is able to trial this on an existing body of TSQL if they have access to it.

**Comment by daviewales** [comment]
I've had a stab at a very simple fix.
I might need some guidance if I need to add tests, etc.

Also, I've tweaked the regex so it detects square brackets as quotes for all dialects, but I'm guessing this should only be enabled for tsql?
